ambari-server/src/main/java/org/apache/ambari/server/security/encryption/PasswordPropertiesEncryptor.java

```diff
@@ -20,45 +20,61 @@
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 
 import org.apache.ambari.server.AmbariException;
 import org.apache.ambari.server.AmbariRuntimeException;
+import org.apache.ambari.server.events.StackUpgradeFinishEvent;
+import org.apache.ambari.server.events.publishers.AmbariEventPublisher;
 import org.apache.ambari.server.state.Cluster;
-import org.apache.ambari.server.state.Clusters;
 import org.apache.ambari.server.state.PropertyInfo.PropertyType;
 import org.apache.ambari.server.utils.TextEncoding;
+import org.apache.commons.collections.CollectionUtils;
 
+import com.google.common.eventbus.Subscribe;
 import com.google.inject.Inject;
 import com.google.inject.Singleton;
 
 /**
- * {@link Encryptor} implementation for encrypting/decrypting PASSWORD type properties
- *
+ * {@link Encryptor} implementation for encrypting/decrypting PASSWORD type
+ * properties
  */
-    
+
 @Singleton
 public class PasswordPropertiesEncryptor implements Encryptor<Map<String, String>> {
 
   private static final String ENCRYPTED_PROPERTY_PREFIX = "${enc=aes256_hex, value=";
   private static final String ENCRYPTED_PROPERTY_SCHEME = ENCRYPTED_PROPERTY_PREFIX + "%s}";
 
   private final EncryptionService encryptionService;
-  private final Map<Long, Map<String, Set<String>>> clusterPasswordProperties = new HashMap<>();
+  private final Map<Long, Map<String, Set<String>>> clusterPasswordProperties = new ConcurrentHashMap<>();
 
   @Inject
-  public PasswordPropertiesEncryptor(Clusters clusters, EncryptionService encryptionService) {
+  public PasswordPropertiesEncryptor(EncryptionService encryptionService, AmbariEventPublisher ambariEventPublisher) {
     this.encryptionService = encryptionService;
+    ambariEventPublisher.register(this);
   }
 
+  /**
+   * It is a must that any caller of this method adds the following additional
+   * information on the following indexes in <code>additionalInfo</code> with the
+   * following types:
+   * <ul>
+   * <li>[0] - <code>org.apache.ambari.server.state.Cluster</code> (the cluster
+   * whose properties should be scanned for PASSWORD properties)</li>
+   * <li>[1] - <code>java.lang.String</code> (to keep properties only that match
+   * this configuration type )</li>
+   * </ul>
+   */
   @Override
-  public Map<String, String> encryptSensitiveData(Map<String, String> properties, Object... additionaKeys) {
+  public Map<String, String> encryptSensitiveData(Map<String, String> properties, Object... additionaInfo) {
```

Comment:
Any reason why we pass these arguments as a Object[] ? I see it comes from the interface, but the caller still needs to know which is the underlying implementation in use to fill these params correctly.

Reply:
But the caller should not be down casted to the implementation.